src/GraphQL/Instrumentation/GraphQLTelemetryProvider.cs

```diff
@@ -0,0 +1,140 @@
+    /// Initializes a new instance with the specified options.
+    /// </summary>
+    public GraphQLTelemetryProvider(GraphQLTelemetryOptions options)
+    {
+        _telemetryOptions = options;
+    }
+
+    /// <inheritdoc/>
+    public virtual float SortOrder => GraphQLBuilderExtensions.SORT_ORDER_CONFIGURATION;
+
+    /// <inheritdoc/>
+    public virtual async Task<ExecutionResult> ExecuteAsync(ExecutionOptions options, ExecutionDelegate next)
+    {
+        // start the Activity
+        using var activity = await StartActivityAsync(options).ConfigureAwait(false);
+
+        // if no event listeners, do not record any telemetry
+        if (activity == null)
+            return await next(options).ConfigureAwait(false);
+
+        // record the requested operation name and optionally the GraphQL document
+        await SetInitialTagsAsync(activity, options).ConfigureAwait(false);
+
+        // record the operation type and the operation name (which may be specified within the
+        // document even if not specified in the request)
+        options.Listeners.Add(new TelemetryListener(this, activity, options));
+
+        // execute the request
+        var result = await next(options).ConfigureAwait(false);
+
+        // record the status
+        await SetResultTagsAsync(activity, options, result).ConfigureAwait(false);
+
+        // return the result
+        return result;
+    }
+
+    /// <summary>
+    /// Creates an <see cref="Activity"/> for the specified <see cref="ExecutionOptions"/> and starts it.
+    /// </summary>
+    protected virtual ValueTask<Activity?> StartActivityAsync(ExecutionOptions options)
+        => new(_activitySource.StartActivity(ACTIVITY_OPERATION_NAME));
+
+    /// <summary>
+    /// Sets the <see cref="Activity"/> tags based on the specified <see cref="ExecutionOptions"/>.
+    /// The default implementation sets the <c>graphql.operation.name</c> and <c>graphql.document</c> tags.
+    /// </summary>
+    protected virtual Task SetInitialTagsAsync(Activity activity, ExecutionOptions options)
+    {
+        activity.SetTag("graphql.operation.name", options.OperationName);
+        if (_telemetryOptions.RecordDocument && activity.IsAllDataRequested)
+            activity.SetTag("graphql.document", options.Query);
+        return Task.CompletedTask;
+    }
+
+    /// <summary>
+    /// Sets the <see cref="Activity"/> tags based on the specified <see cref="GraphQLOperationDefinition"/> and other values.
+    /// The default implementation sets the <c>graphql.operation.name</c> and <c>graphql.operation.type</c> tags, and sets the
+    /// <see cref="Activity.DisplayName"/> property based on the operation name and type.
+    /// </summary>
+    protected virtual Task SetOperationTagsAsync(Activity activity, ExecutionOptions options, ISchema schema, GraphQLDocument document, GraphQLOperationDefinition operation)
+    {
+        var operationType = operation.Operation.ToString().ToLowerInvariant();
```

Comment:
I know that `enum.ToString()` has already been optimized to not allocate strings each call but this (among `ToLowerInvariant` call) makes me doubt. Maybe use 3-line local helper here with switch expression returning static strings as is?

Reply:
I agree.

Reply:
Done